highlighters/main/main-highlighter.zsh

```diff
@@ -165,88 +147,72 @@ _zsh_highlight_main__type() {
   # ### $aliases_allowed, on the assumption that aliases are the common case.
   integer may_cache=1
 
-  # Cache lookup
-  if (( $+_zsh_highlight_main__command_type_cache )); then
-    REPLY=$_zsh_highlight_main__command_type_cache[(e)$1]
-    if [[ -n "$REPLY" ]]; then
-      return
-    fi
-  fi
-
   # Main logic
-  if (( $#options_to_set )); then
-    setopt localoptions $options_to_set;
-  fi
   unset REPLY
   if zmodload -e zsh/parameter; then
-    if (( $+aliases[(e)$1] )); then
+    if (( $+aliases[$1] )); then
       may_cache=0
     fi
-    if (( ${+galiases[(e)$1]} )) && (( aliases_allowed )); then
+    if (( ${+galiases[$1]} )) && (( aliases_allowed )); then
       REPLY='global alias'
-    elif (( $+aliases[(e)$1] )) && (( aliases_allowed )); then
+    elif (( $+aliases[$1] )) && (( aliases_allowed )); then
       REPLY=alias
-    elif [[ $1 == *.* && -n ${1%.*} ]] && (( $+saliases[(e)${1##*.}] )); then
+    elif [[ $1 == *.* && -n ${1%.*} ]] && (( $+saliases[${1##*.}] )); then
       REPLY='suffix alias'
     elif (( $reswords[(Ie)$1] )); then
       REPLY=reserved
-    elif (( $+functions[(e)$1] )); then
+    elif (( $+functions[$1] )); then
       REPLY=function
-    elif (( $+builtins[(e)$1] )); then
+    elif (( $+builtins[$1] )); then
       REPLY=builtin
-    elif (( $+commands[(e)$1] )); then
+    elif (( $+commands[$1] )); then
       REPLY=command
-    # None of the special hashes had a match, so fall back to 'type -w', for
-    # forward compatibility with future versions of zsh that may add new command
-    # types.
-    #
-    # zsh 5.2 and older have a bug whereby running 'type -w ./sudo' implicitly
-    # runs 'hash ./sudo=/usr/local/bin/./sudo' (assuming /usr/local/bin/sudo
-    # exists and is in $PATH).  Avoid triggering the bug, at the expense of
-    # falling through to the $() below, incurring a fork.  (Issue #354.)
-    #
-    # The first disjunct mimics the isrelative() C call from the zsh bug.
-    elif {  [[ $1 != */* ]] || is-at-least 5.3 } &&
-         # Add a subshell to avoid a zsh upstream bug; see issue #606.
-         # ### Remove the subshell when we stop supporting zsh 5.7.1 (I assume 5.8 will have the bugfix).
-         ! (builtin type -w -- "$1") >/dev/null 2>&1; then
-      REPLY=none
+    # ZSH_VERSION >= 5.1 allows the use of #q. ZSH_VERSION <= 5.8 allows skipping
+    # 'type -w' calls that are necessary for forward compatibility.
+    elif [[ $ZSH_VERSION == 5.<1-8>(|.*) ]]; then
+      if [[ $1 == */* && -n $1(#qN-.*) ||
+            $1 == [^/]*/* && $zsyh_user_options[pathdirs] == on && -n ${^path}/$1(#q-N.*) ]]; then
+        REPLY=command
+      elif (( _zsh_highlight_main__rehash )); then
+        builtin rehash
```

Comment:
> My goal was faster highlighting of partially typed commands. For example, when I type `systemctl status` one character at a time, `master` forks twice on every character until I get to `systemctl`. Two forks (or even just one) on every keystroke is something I'd like to avoid.

Makes sense.

Previous discussions about the "command word whilst it's being typed" case: #148, #244.

> Are there options that make the second `rsync` command fail with "command not found"?

No, sorry, I must've been misremembering.  I stand corrected.

> ```zsh
> % sudo docker run -e TERM -e LC_ALL=C.UTF-8 -it --rm zshusers/zsh:5.8
> # rsync
> zsh: command not found: rsync
> # apt-get update && apt-get install -y rsync
> # rsync
> rsync  version 3.1.2  protocol version 31
> # print $+commands[rsync]
> 0
> ```
> 
> _Tangent: Seems like the meaning of `hash_cmds` option is reversed. If the commands above are executed with `no_hash_cmds`, `$+commands[rsync]` ends up set to `1` at the end._

I see that too:

```zsh
$ sudo chmod -x /usr/local/bin/caste
$ zsh -f -o nohashcmds
% sudo chmod +x /usr/local/bin/caste
% caste &>/dev/null 
% echo $+commands[caste] 
1
```

But checking another way, I don't:

```zsh
$ sudo chmod -x /usr/local/bin/caste
$ zsh -f -o nohashcmds
% sudo chmod +x /usr/local/bin/caste
% caste &>/dev/null
% hash | grep caste
%
```

I think the `1` output in the NO_HASH_CMDS case is because of the implicit `cmdnamtab->filltable()` call in the getter when the HASH_LIST_ALL option is set: https://github.com/zsh-users/zsh/blob/zsh-5.8/Src/Modules/parameter.c#L217-L221.  However, I can't explain the `0` output in the "default option settings" case.  Isn't it a bug?

I don't have a strong opinion regarding whether or not HASH_CMDS should or shouldn't affect ${commands}.

> I've added a test that verifies that `rehash` is not called (`no-rehash.zsh`). It fails with the version of this PR that invokes `rehash`, and succeeds in `master`.
>
> I've also added a test (`removed-command.zsh`) that verifies highlighting of a newly uninstalled command (e.g., after `apt remove rsync`). Since the command cannot be invoked, it should be highlighted as "unknown-token" (red). This test fails on `master` and the version of this PR that you've looked at.

Thanks for the additional tests and bugfixes!

> I believe it's desirable for `foo` to be highlighted in red here. Let me know if I misunderstood the expected behavior of zsyh.

That's an edge case indeed, but I agree it should be highlighted in red.

> I've changed `main-highlighter.zsh` so that all tests pass. It's still as fast as in the very first version of the PR. No forks when using a recent-enough version of zsh. Note that this code now highlights some tokens as "hashed-command" that were previously (mistakenly, I think) highlighted as "command".

Thanks.  Looks good, but a few comments:

1. The log message of the latest commit (6a210f1408101fd075f52cc48b04a6ab44a661ac) isn't clear about _which_ corner cases were fixed.  Is it just "the cases that tests are being added for"?

2. `There is one case where the following logic incorrectly sets` ← could you please either add an XFail test, or make this a `### TODO:` comment, or both?

3. Please explain the magic number `5.8` in that comment.  (I guess it should just say that that's the latest zsh release at the time of writing.)

4. `-n $1(#q-.*N)` seems a little roundabout: why not `[[ -x $1 ]]`?  (The answer should be in a code comment, please, rather than here.)

5. Should `_zsh_highlight_main__command_type_cache` be emptied when we notice `$PATH` or `${zsyh_user_options[pathdirs]}` has changed?  (Preëxisting issue, at least for PATH_DIRS)

6. Now the PR adds `_zsh_highlight_main__rehash` and then removes it again.  What's your preference — to keep it like that, or to rewrite history (after review finishes, before merging) to avoid the intermediate state?  (Honest question.)

7. What's the difference between the `hashed-command` test in master and the `invalid-hashed-command` test in the PR?  Seems like they're identical, except that one is XFailing and one is passing?  The latest commit seems to have essentially repurposed `hashed-command` to test a different scenario than it tests in master, relegating the scenario tested in master to another file, and I don't see why.

8. `no-rehash`: Cherry-picked in 62c5575848f1.  Thanks!

9. Changing `hash sudo=false` is correct, but isn't covered by the log message.

Thanks for all the work so far!

Reply:
I saw your comments and I appreciate them. I should be able to address them early next week.

Reply:
> I can't explain the `0` output in the "default option settings" case. Isn't it a bug?

Looks like a bug.

> I don't have a strong opinion regarding whether or not HASH_CMDS should or shouldn't affect ${commands}.

My expectation is that `$commands` contains hashed commands and only hashed commands. So if `HASH_CMDS` causes an invoked command to be hashed, `$commands` should reflect that. I don't have a strong support for this expectation, it's just how I _thought_ it works.

> 1. The log message of the latest commit ([6a210f1](https://github.com/zsh-users/zsh-syntax-highlighting/commit/6a210f1408101fd075f52cc48b04a6ab44a661ac)) isn't clear about _which_ corner cases were fixed.

See below.

> Is it just "the cases that tests are being added for"?

Yes.

> 2. `There is one case where the following logic incorrectly sets` ← could you please either add an XFail test, or make this a `### TODO:` comment, or both?

Done. Added `highlighters/main/test-data/ambiguous-hashed-command.zsh`.

Note that this isn't a regression. The PR fixes most misclassifications of hashed commands as plain commands but it doesn't fix all of them.

> 3. Please explain the magic number `5.8` in that comment.  (I guess it should just say that that's the latest zsh release at the time of writing.)

Done.

> 4. `-n $1(#q-.*N)` seems a little roundabout: why not `[[ -x $1 ]]`?  (The answer should be in a code comment, please, rather than here.)

Added a comment.

```zsh
# [[ -n $1(#q-.*N) ]] is a faster version of [[ -f $1 && -x $1 ]].
```

The difference in performance is greatest on filesystems with slow `stat()`. Another reason I've written it this way is consistency with the code a few lines below.

> 5. Should `_zsh_highlight_main__command_type_cache` be emptied when we notice `$PATH` or `${zsyh_user_options[pathdirs]}` has changed?  (Preëxisting issue, at least for PATH_DIRS)

`_zsh_highlight_main__command_type_cache` is emptied on `precmd`. I think this is reasonable. (I think computing `zsyh_user_options` only once per command would also be reasonable but it makes no noticeable difference on performance.)

> 6. Now the PR adds `_zsh_highlight_main__rehash` and then removes it again.  What's your preference — to keep it like that, or to rewrite history (after review finishes, before merging) to avoid the intermediate state?  (Honest question.)

I don't know the standard PR methodology or even if there is one. I usually do it like this:

1. Work on code changes locally. Commit when I feel like.
2. Once done, rebase all commits. Split changes into reasonable self-contained chunks so that commits can be reviewed one-by-one.
3. Once the reviewer posts their first comments, avoid force pushes. Additional changes requested by the reviewer must go in separate commits.
4. When the reviewer is happy with the code changes, ask them if they prefer the PR to be rebased and whether they have a preference for how it should be split into commits.

I think we are currently at 3 -- discussing the code changes and making changes to it. If you are OK with the current state of the code, let me know whether you want me to rebase the PR into one commit with a long description or a handful of commits with somewhat shorter descriptions. The former is obviously easier for me but I'm fine with the latter, too.

> 7. What's the difference between the `hashed-command` test in master and the `invalid-hashed-command` test in the PR?  Seems like they're identical, except that one is XFailing and one is passing?  The latest commit seems to have essentially repurposed `hashed-command` to test a different scenario than it tests in master, relegating the scenario tested in master to another file, and I don't see why.

It's a trade-off between the simplicity of the final version of the code and the size of the code delta to reach it. I wanted to have two tests for hashed commands: one where the command is valid and another where it's invalid. `hashed-command` and `invalid-hashed-command` seemed like good names for these tests, so I grabbed them.

That said, I have a very weak preference here, so if you'd like me to change anything I'll be happy to oblige.

> 9. Changing `hash sudo=false` is correct, but isn't covered by the log message.

Indeed. I realize that commit messages are not up to z-sy-h standards. I wanted to invest only as much time in them as would be necessary for you to review the code. I can write higher-quality commit messages once you are OK with the proposed code changes.

> Thanks for all the work so far!

Thanks for the review and great feedback!

Please don't hesitate using more direct language to request changes from me. "Add a comment here", "rename this test", "merge these commits", etc. I'm happy to comply as soon as I understand what's required. By proposing these code changes I'm effectively asking you to maintain them going forward, so it's only reasonable that you request compliance with your own standards.